### PR TITLE
Doctrine provider

### PR DESCRIPTION
**Pull request by ghost**
Hi

As discussed a while ago here's an implementation of a Doctrine provider for qpush.  The idea is to have a table 'uecode_qpush_message' that holds all the messages for all the queues for the provider.  Each message has a queue_name and a boolean to mark it as delivered or not.  The message gets serialised by doctrine and then stored/retrieved from the table.  Messages are not deleted from the table apart from when the destroy method is called.  So the message table has a history of all the message sent along with created and updated times.

To use - examples

config.yml

> uecode_qpush:

```
providers:
    doctrine_provider:
        driver: doctrine  
queues:
    modified_products:
        provider: doctrine_provider
```

You can optionally pass in an entity manager for doctrine.  Otherwise it defaults to doctrine.orm.default_entity_manager

entity_manager:  my.entity.manager

Add this to composer.json

```
    "doctrine/orm":           "^2.4.8",
    "stof/doctrine-extensions-bundle": "^1.2"
```

To create the table use  console doctrine:schema:update

Other than that usage is as per the other providers.

### CODE REVIEW

File: src/Provider/DoctrineProvider.php [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+     * Get repository
+     * 
+     * @return array
+     */
+    public function getRepository()
+    {
+        if (!$this->repository) {
+            return;
+        }
+        return $this->repository;
```
Comment (kmfk) [inline src/Provider/DoctrineProvider.php]: Line break before the return statement

File: src/Provider/DoctrineProvider.php [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+     */
+    public function receive(array $options = [])
+    {
+        if (!$this->em) {
+            return [];
+        }
+
+        $doctrineMessages = $this->repository->findBy(
+                [
+            'delivered' => false,
```
Comment (kmfk) [inline src/Provider/DoctrineProvider.php]: Any chance we can clean up this tab spacing here

File: src/Provider/DoctrineProvider.php [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+     *
+     * @return bool
+     */
+    public function destroy()
+    {
+        $qb = $this->repository->createQueryBuilder('dm');
+        $qb->delete();
+        $qb->where('dm.queue = :queue');
+        $qb->setParameter('queue', $this->name);
+        $qb->getQuery()->getResult();
```
Comment (kmfk) [inline src/Provider/DoctrineProvider.php]: Don't remember off the top of my head what `getResult()` should return for the delete - but we're type hinting that we return a `bool`, but have no return statement.

**Comment by kmfk** [comment]
Hey @talisman-innovations! I added a couple minor CS type comments, otherwise, this is looking pretty good.

**Comment by talisman-innovations** [comment]
Hi @kmfk - no problem - I think I got them all, plus added an implementation of the create function for good measure.  Regards Steve

**Comment by talisman-innovations** [comment]
For a use case we have I'm thinking of implementing a function to allow for redelivery of a message.  As the messages sit in the DB it's trivial to implement.  Any issues with adding that?

Tt could go in the AbstractProvider if other providers offer that feature.  
Any thoughts on naming - maybe republish(id)?

**Comment by kmfk** [comment]
@talisman-innovations I have no memory of why this would still be open for so long.  Looking it over, this looks good to me, but wanted to confirm with you before merging.

**Comment by talisman-innovations** [comment]
Hi - yes feel free to merge.  Thanks.

**Comment by kmfk** [comment]
Will do.  Currently need to fix an issue with PHPUnit and Symfony ^3.2 to fix the failing builds.  Will get this in afterward.

**Comment by drzraf** [comment]
This provider is not documented.
In production environment, even if this provider is NOT used, `cache:clear` triggers:

```
In AnnotationException.php line 54:
  [Semantical Error] The annotation "@Gedmo\Mapping\Annotation\Timestampable" in property Uecode\Bundle\QPushBundle\Entity\DoctrineMessage::$created does not exist, or could not be auto-loaded.  
```

unless `stof/doctrine-extensions-bundle` is installed.

**Pull request merged**